AP_Scripting: add support for I2C bus

IamPete1 opened the pull request:
This adds support for scripting to access the I2C bus. There is a device scan example that returns the address of any device found and also a 'library' for NAU7805 (sparkfun Qwiic Scale) load cell amplifier. This library is begging for 'requires' support.

I have had to add a new way to access I2C devices without using 'own pointer' (cant say I have a 100% grasp on what 'own pointer' does but I suspect we are using it more than necessary?) 

I have manually edited the generated bindings to get this to work, If the general method is approved of we can either alter the generator or move them to manual bindings. 

This only adds support for reading or writing a single byte at a time, doing more than this is much more complicated, but I think everything can still work like this just might not be the most efficient, but its in scripting anyway so I don't think its a disaster.

Only tested so far with one thing on the bus, will also test AP talking to something else at the same time.

tridge commented:
I think this can be done without the HAL changes. Have a look at this code for an example:
https://github.com/ardupilot/ardupilot/blob/master/libraries/GCS_MAVLink/GCS_DeviceOp.cpp#L29

IamPete1 commented:
@tridge so we get a new device for each transaction? I think that would work, any reason that is better?

IamPete1 commented:
use std::move to deal with own_ptr

tridge commented:
needs to use std::move for ownptr, and allow for more than 1 byte read/write

IamPete1 commented:
Re-based and moved to manual bindings, spent a significant amount of time trying to get std::move() to work with no luck. Still only one byte at a time, but so is the current UART implementation, I think that can wait for a future improvement.

tridge commented:
I'm tempted to remove the use of ownptr in ArduPilot. I've found it a major pain too

IamPete1 commented:
rebased, added several features to the generator such that these bindings are now auto generated. still does not fetch more than one byte at a time, but that limitation is there for UART too, and still does not use `std::move` because I have not been able to move into a lua userdata object. 

Generator improvements: 

- support literal singletons, these are used directly rather than fetching a singleton, for `hal` function in this example
- support sanitized names and alias for methods. In combination this allows to reach into sub clases, for example we could now do a method called `copter.mode_zigzag.save_or_move_to_destination` and alias it to `zigzag_save`. This also allows us to strip out all the helper functions we have added so scripting can see a more friendly name, we swap the method to the full name and alias it to the old helper name.
- support to reference access for functions that require pointers. 
- support for semaphore pointers, this let us use `get_semaphore()->take_blocking();` rather than `get_semaphore().take_blocking();`

WickedShell commented on libraries/AP_HAL/I2CDevice.h:
```diff
@@ -76,6 +76,11 @@ class I2CDeviceManager {
                                                  bool use_smbus = false,
                                                  uint32_t timeout_ms=4) = 0;
+
+    virtual AP_HAL::I2CDevice * get_device_direct(uint8_t bus, uint8_t address,
```
Needs some HAL_Linux/HAL_SITL implementations. I think @peterbarker has enough going that at a minimum SITL should be added and testable these days as a blocker, before merge. Linux may be a separate debate.

IamPete1 replied:
removed this funciton

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -328,3 +328,15 @@ singleton AP_MotorsMatrix_6DoF_Scripting depends APM_BUILD_TYPE(APM_BUILD_ArduCo
+
+singleton hal.i2c_mgr alias i2c
+singleton hal.i2c_mgr literal
+singleton hal.i2c_mgr method get_device_direct AP_HAL::I2CDevice uint8_t 0 4 uint8_t 0 UINT8_MAX
```
4 should probably be a define we get from the HAL somewhere for the number of buses possible. Also probably worth considering that you almost guranteed need the clock speed, and allowing using SMBus would actually be valuable as a flag.

IamPete1 replied:
I don't think there is a unified define for this over all the HALs

IamPete1 replied:
added clock speed and SMBus flag

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -1445,8 +1504,14 @@ void emit_userdata_method(const struct userdata *data, const struct method *meth
+
   if (data->flags & UD_FLAG_SEMAPHORE) {
-    fprintf(source, "    ud->get_semaphore().take_blocking();\n");
+    fprintf(source, "    %s->get_semaphore().take_blocking();\n", ud_name);
+  }
+
+  if (data->flags & UD_FLAG_SEMAPHORE_POINTER) {
+    fprintf(source, "    %s->get_semaphore()->take_blocking();\n", ud_name);
```
```suggestion
    fprintf(source, "    %s->get_semaphore().take_blocking();\n", ud_name);
  } else if (data->flags & UD_FLAG_SEMAPHORE_POINTER) {
    fprintf(source, "    %s->get_semaphore()->take_blocking();\n", ud_name);
```

Let's make it more obvious that these cases are mutual exclusive.

IamPete1 replied:
fixed

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -1573,7 +1638,11 @@ void emit_userdata_method(const struct userdata *data, const struct method *meth
 
   if (data->flags & UD_FLAG_SEMAPHORE) {
-    fprintf(source, "    ud->get_semaphore().give();\n");
+    fprintf(source, "    %s->get_semaphore().give();\n", ud_name);
+  }
+
+  if (data->flags & UD_FLAG_SEMAPHORE_POINTER) {
+    fprintf(source, "    %s->get_semaphore()->give();\n", ud_name);
```
```suggestion
    fprintf(source, "    %s->get_semaphore().give();\n", ud_name);
  } else if (data->flags & UD_FLAG_SEMAPHORE_POINTER) {
    fprintf(source, "    %s->get_semaphore()->give();\n", ud_name);
```

Same as above.

IamPete1 replied:
fixed

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -328,3 +328,15 @@ singleton AP_MotorsMatrix_6DoF_Scripting depends APM_BUILD_TYPE(APM_BUILD_ArduCo
+
+include AP_HAL/I2CDevice.h
+ap_object AP_HAL::I2CDevice semaphore-pointer
+ap_object AP_HAL::I2CDevice method set_retries void uint8_t 1 20
```
Why can't retries be 0?

IamPete1 replied:
it can, fixed

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -328,3 +328,15 @@ singleton AP_MotorsMatrix_6DoF_Scripting depends APM_BUILD_TYPE(APM_BUILD_ArduCo
+ap_object AP_HAL::I2CDevice semaphore-pointer
+ap_object AP_HAL::I2CDevice method set_retries void uint8_t 1 20
+ap_object AP_HAL::I2CDevice method write_register boolean uint8_t 0 UINT8_MAX uint8_t 0 UINT8_MAX
+ap_object AP_HAL::I2CDevice method read_registers boolean uint8_t 0 UINT8_MAX &uint8_t'Null 1'literal
```
This is a case where we probably need a more explicit helper for scripting manually added. Most useful I2C devices return more then a single byte from the register (for example you can't implement a SMBus driver without reading multi byte registers).

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -377,6 +383,42 @@ void sanatize_name(char **dest, char *src) {
     position = strchr(position, ':');
   }
+
+  position = strchr(*dest, '.');
+  while (position) {
+    *position = '_';
+    position = strchr(position, '.');
+  }
```
Let's break this out as a core helper of "sanatize_character()" or something along those lines (I'm bad at naming) that simply takes a character to replace everywhere and then call that. This is a lot of duplication in here now, and it's more likely to have a bug in the future. (Also once it's a helper the whole thing can be expressed as a for each char in a string of chars to sanatize)

IamPete1 replied:
fixed

WickedShell reviewed (changes requested):
Not a complete review as I ran out of time, but I'm a bit alarmed at allowing the user to allocate as many I2C devices as they want outside of the scripting memory sandbox. This gives them unfettered access to system memory, which scripting is intentionally not allowing. I think the I2C interface needs a wrapper that sets a hard limit on how many I2C devices you are ever allowed to fetch, and that needs to persist across scripting restarts, or other problems may occur in the flight controller.

tridge commented:
this needs an example script

IamPete1 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -901,15 +958,17 @@ void handle_ap_object(void) {
   // check that we didn't just add 2 singleton flags
-  if ((node->flags & UD_FLAG_SEMAPHORE) && (node->flags & UD_FLAG_SCHEDULER_SEMAPHORE)) {
-    error(ERROR_SINGLETON, "Taking both a library semaphore and scheduler semaphore is prohibited");
+  if ((node->flags & UD_FLAG_SEMAPHORE) && (node->flags & UD_FLAG_SCHEDULER_SEMAPHORE) && (node->flags & UD_FLAG_SEMAPHORE_POINTER)) {
```
fixed

IamPete1 commented on libraries/AP_HAL_ChibiOS/I2CDevice.cpp:
```diff
@@ -367,6 +367,19 @@ I2CDeviceManager::get_device(uint8_t bus, uint8_t address,
+    if (bus >= ARRAY_SIZE(I2CD)) {
+        return nullptr;
+    }
+    auto dev = new I2CDevice(bus, address, bus_clock, use_smbus, timeout_ms);
```
fixed this with new helper class, now only allowed up to 4 devices per boot

IamPete1 commented:
this now includes a `scripting_helper_functions` class, this only lets scripting have upto 4 i2c devices per boot. This class can have similar functions added to it, for example for analog ins from https://github.com/ArduPilot/ardupilot/pull/16449

IamPete1 commented:
now is done a custom binding rather than helper class, also allows defaults to be used for clock speed and SMBus flag.

rmackay9 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -369,14 +375,26 @@ struct userdata {
   }
+}
+
+void sanatize_name(char **dest, char *src) {
```
"sanatize" should be "sanitise" or "sanitize".  we normally use the british english spelling which would be "sanitise" but either is ok.

IamPete1 replied:
that is a existing issue, I have fixed the name on the new `sanitize_character` function.

rmackay9 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -528,6 +550,8 @@ int parse_type(struct type *type, const uint32_t restrictions, enum range_check_
   } else if (type->type == TYPE_LITERAL) {
       string_copy(&(type->data.literal), data_type);
+  } else if (strcmp(data_type, keyword_alias) == 0) {
+    error(ERROR_USERDATA, "Cant add alias to unknown function");
```
tiny whitespace issue here I think.  just indent a couple of characters.

WickedShell replied:
It's actually not a whitespace issue here, the whole file was 2 space indentation. Honestly the 4 spaces up on line 525 is the bigger change.

IamPete1 replied:
I have fixed the indentation on the above line.

tridge commented:
I really like this feature. ping @WickedShell

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -901,15 +936,20 @@ void handle_ap_object(void) {
   // check that we didn't just add 2 singleton flags
-  if ((node->flags & UD_FLAG_SEMAPHORE) && (node->flags & UD_FLAG_SCHEDULER_SEMAPHORE)) {
-    error(ERROR_SINGLETON, "Taking both a library semaphore and scheduler semaphore is prohibited");
+  int SEMAPHORE = (node->flags & UD_FLAG_SEMAPHORE)?1:0;
```
lower case? I'm just not sure why this is caps locked.

IamPete1 replied:
Its almost as though I copied and passed it from the enum..... fixed

WickedShell commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -45,10 +47,56 @@ static int lua_micros(lua_State *L) {
+    }
+
+    AP::scripting()->_i2c_dev[AP::scripting()->num_i2c_devices] = new AP_HAL::OwnPtr<AP_HAL::I2CDevice>;
+    *AP::scripting()->_i2c_dev[AP::scripting()->num_i2c_devices] = std::move(hal.i2c_mgr->get_device(bus, address, bus_clock, use_smbus));
```
The device you get could be `nullptr` you need to actually check that it isn't a nullptr here. If it is we should probably be declaring an error.

IamPete1 replied:
added a check for both the own pointer being null and the pointer in the own pointer being null

WickedShell replied:
You have to check that the OwnPtr isn't null here, not after you've dereferenced it

IamPete1 replied:
fixed

WickedShell commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -45,10 +47,56 @@ static int lua_micros(lua_State *L) {
 {
     {"millis", lua_millis},
     {"micros", lua_micros},
+    {"get_i2c_device", lua_get_i2c_device},
```
This would probably be better off in a table called "i2c" and then just have the method be called "get_device", but it's not critical or a blocker.

IamPete1 replied:
moved

WickedShell commented on libraries/AP_Scripting/AP_Scripting.h:
```diff
@@ -20,6 +20,10 @@
 #include <AP_Filesystem/AP_Filesystem.h>
+#include <AP_HAL/I2CDevice.h>
+
+#define SCRIPTING_MAX_NUM_I2C_DEVICE 4
```
Maybe add a `#ifndef` check before this, that way a hwdef can override it?

IamPete1 replied:
done

WickedShell commented on libraries/AP_Scripting/lua_bindings.cpp:
```diff
@@ -45,10 +47,56 @@ static int lua_micros(lua_State *L) {
+        }
+    }
+
+    if (AP::scripting()->num_i2c_devices >= SCRIPTING_MAX_NUM_I2C_DEVICE) {
```
```suggestion
   static_assert(MAX_NUM_I2C_DEVICE >= 0, "There cannot be a negative number of I2C devices");
    if (num_i2c_devices >= MAX_NUM_I2C_DEVICE) {
```

IamPete1 replied:
done

WickedShell reviewed (changes requested):
The failure case on get_device needs some more attention first. You should be able to test this by just tweaking the bus number to allow a larger input, and requesting a nonsensical bus.

@tridge do you think there is any point in sanity checking the I2C bus parameter (past being a uint8_t)? I know we just had lots of debates on the num instances elsewhere. If @IamPete1 is correct that there isn't a HAL define, then the hard coded 4 is probably just a ticking time bomb of annoyance for the future. (By the same token I haven't managed to check if all the HAL functions audit the bus safely)

IamPete1 commented on libraries/AP_Scripting/AP_Scripting_helpers.h:
```diff
@@ -20,3 +21,32 @@ class Parameter
     AP_Param *vp;
 };
 
+#define MAX_NUM_I2C_DEVICE 4
```
done

IamPete1 commented on libraries/AP_Scripting/AP_Scripting_helpers.cpp:
```diff
@@ -93,3 +96,20 @@ bool Parameter::set_and_save(float value)
 
+AP_HAL::I2CDevice* scripting_helper_functions::get_i2c_device(uint8_t bus, uint8_t address, uint32_t bus_clock, bool use_smbus)
+{
+    if (num_i2c_devices >= MAX_NUM_I2C_DEVICE) {
```
added assert

WickedShell commented:
I think I'm pretty happy with the proposed changes, I'd just really appreciate someone else doing a review of the manual I2C binding for any other edge cases I missed (I was a slow reviewer to catch a couple of things there, and I'm now worried about anything else obvious I missed in the I2C binding)

IamPete1 commented:
rebased
